### PR TITLE
DOC: update ufunc tutorials to use setuptools

### PR DESCRIPTION
**Pull request by JuliaPoo**
See #22027. Currently ufunc tutorials use depreciated distutils which is removed in Python 3.12. In addition, I've updated the sample output and fixed a mistake in the last example.

### CODE REVIEW

**Comment by mattip** [comment]
I wonder if we could convert this tutorial (and the others like the [cython one](https://numpy.org/devdocs/user/c-info.python-as-glue.html#cython) to use meson. The [extending random examples](https://numpy.org/devdocs/user/c-info.python-as-glue.html#cython) have a meson.build file, but the documentation is a little sparse. The [f2py migration guide](https://numpy.org/devdocs/f2py/buildtools/distutils-to-meson.html) is a little better, but could also use some fleshing out.

**Comment by rgommers** [comment]
Thanks @JuliaPoo!

> I wonder if we could convert this tutorial (and the others like the [cython one](https://numpy.org/devdocs/user/c-info.python-as-glue.html#cython) to use meson.

I think that is a much larger exercise, and best done separately if someone has energy for that. `setuptools` is still a valid and common way of building extensions against `numpy`, so no need to get rid of that. Migrating from `numpy.distutils` to `setuptools` is a nice improvement in itself, that makes things work on Python 3.12. In an ideal world, we'd have examples for all the popular build backends (at least `meson-python`, `scikit-build-core` and `setuptools`) everywhere - but that's quite a bit of work.

File: doc/source/user/c-info.ufunc-tutorial.rst [hunk 1]
```diff
@@ -182,21 +182,20 @@ site-packages directory.
...
         '''
 
+        from setuptools import setup, Extension
+        import numpy as np
 
-        from distutils.core import setup, Extension
-
-        module1 = Extension('spam', sources=['spammodule.c'],
-                                include_dirs=['/usr/local/lib'])
+        module1 = Extension('spam',
```
Comment (mhvk) [inline doc/source/user/c-info.ufunc-tutorial.rst:193]: Maybe bring on one line for easier copy & pasting.

File: doc/source/user/c-info.ufunc-tutorial.rst [hunk 2]
```diff
@@ -182,21 +182,20 @@ site-packages directory.
...
+                            sources=['spammodule.c'])
 
         setup(name = 'spam',
-                version='1.0',
-                description='This is my spam package',
-                ext_modules = [module1])
+            version='1.0',
```
Comment (mhvk) [inline doc/source/user/c-info.ufunc-tutorial.rst:197]: Align properly (was bad before, might as well correct!)

File: doc/source/user/c-info.ufunc-tutorial.rst [hunk 3]
```diff
@@ -182,21 +182,20 @@ site-packages directory.
             $python setup.py install
             will install the module in your site-packages file.
 
-            See the distutils section of
-            'Extending and Embedding the Python Interpreter'
-            at docs.python.org for more information.
+            See the setuptools section
+            'Building Extension Modules'
```
Comment (mhvk) [inline doc/source/user/c-info.ufunc-tutorial.rst:186]: probably can fit on previous line.

File: doc/source/user/c-info.ufunc-tutorial.rst [hunk 4]
```diff
@@ -373,33 +372,29 @@ using ``python setup.py build_ext --inplace``.
             $python setup.py install
             will install the module in your site-packages file.
 
-            See the distutils section of
-            'Extending and Embedding the Python Interpreter'
-            at docs.python.org  and the documentation
-            on numpy.distutils for more information.
+            See the setuptools section
```
Comment (mhvk) [inline doc/source/user/c-info.ufunc-tutorial.rst:375]: As above, maybe combine this line with next

File: doc/source/user/c-info.ufunc-tutorial.rst [hunk 5]
```diff
@@ -373,33 +372,29 @@ using ``python setup.py build_ext --inplace``.
...
-                                   parent_package,
-                                   top_path)
-            config.add_extension('npufunc', ['single_type_logit.c'])
+        npufunc = Extension('npufunc',
+                            sources=['single_type_logit.c'],
+                            include_dirs=[get_include()])
 
-            return config
+        setup(name = 'npufunc',
+            version='1.0',
```
Comment (mhvk) [inline doc/source/user/c-info.ufunc-tutorial.rst:388]: Alignment (and maybe just on one line)

File: doc/source/user/c-info.ufunc-tutorial.rst [hunk 6]
```diff
@@ -625,38 +622,34 @@ or installed to site-packages via ``python setup.py install``.
             $python setup.py install
             will install the module in your site-packages file.
 
-            See the distutils section of
-            'Extending and Embedding the Python Interpreter'
-            at docs.python.org  and the documentation
-            on numpy.distutils for more information.
+            See the setuptools section
```
Comment (mhvk) [inline doc/source/user/c-info.ufunc-tutorial.rst:625]: Combine lines

File: doc/source/user/c-info.ufunc-tutorial.rst [hunk 7]
```diff
@@ -625,38 +622,34 @@ or installed to site-packages via ``python setup.py install``.
...
+                            libraries=["npymath"])
 
-            config = Configuration('npufunc_directory',
-                                    parent_package,
-                                    top_path)
-            config.add_extension('npufunc',
-                                    ['multi_type_logit.c'],
-                                    extra_info=info)
+        setup(name = 'npufunc',
+            version='1.0',
```
Comment (mhvk) [inline doc/source/user/c-info.ufunc-tutorial.rst:643]: Alignment/combine lines

**Review by mhvk — COMMENTED** [review]
Thanks, great to have documentation that actually works! And thanks for going through it carefully and ensuring the outputs are actually correct, etc.

Also indeed fine to just use `setuptools` for these examples - others can be added later if need be.

All my comments are nitpicks, of the "please align" and "please combine lines" type.

**Comment by JuliaPoo** [comment]
Okie I've modified the PR to these comments

**Review by mhvk — APPROVED** [review]
Nice, thanks! Looks all OK now, so let's get this in.

**Comment by charris** [comment]
@rgommers already backported.

**Comment by rgommers** [comment]
oops sorry, I missed that. thanks!

**Pull request merged**